File: código-fonte/Python/sequencial/correlacao.py
import numpy as np

def aplicar_correlacao(imagem_entrada : np.ndarray, mascara : np.ndarray) -> np.ndarray:
    altura_imagem, largura_imagem = imagem_entrada.shape
    altura_mascara, largura_mascara = mascara.shape

    centro_largura = largura_mascara // 2
    centro_altura = altura_mascara // 2
    
    altura_mascara_par = (altura_mascara % 2 == 0)
    largura_mascara_par = (largura_mascara % 2 == 0)

    imagem_saida = np.zeros((altura_imagem, largura_imagem), dtype=np.uint8)

    for i in range(altura_imagem):
        for j in range(largura_imagem):
            regiao = imagem_entrada[max(0, i-centro_altura):min(altura_imagem, i+centro_altura if altura_mascara_par else i+centro_altura+1),
                        max(0, j-centro_largura):min(largura_imagem, j+centro_largura if largura_mascara_par else j+centro_largura+1)]
            
            imagem_saida[i, j] = np.sum(regiao * mascara[max(0, centro_altura - i):min(altura_mascara, altura_imagem - (i - centro_altura)),
                                        max(0, centro_largura - j):min(largura_mascara, largura_imagem - (j - centro_largura))]).astype(np.uint8)
    
    return imagem_saida

def gerar_mascara_media(dimensao : int):
    mascara = np.zeros((dimensao, dimensao), dtype=np.float32)
    
    media : float = 1 / float(dimensao * dimensao)
    
    mascara[:,:] = media
    
    return mascara

File: código-fonte/C++/sequencial/correlacao.cpp
#include "correlacao.hpp"

Mat aplicar_correlacao(const Mat &imagem_entrada, const Mat &mascara) {
    const int altura_imagem = imagem_entrada.rows;
    const int largura_imagem = imagem_entrada.cols;

    const int altura_mascara = mascara.rows;
    const int largura_mascara = mascara.cols;

    const int centro_altura = altura_mascara / 2;
    const int centro_largura = largura_mascara / 2;

    Mat imagem_saida = Mat::zeros(altura_imagem, largura_imagem, CV_8UC1); // 8 bits uchar 1 banda

    for (int i = 0; i < altura_imagem; i++)
        for (int j = 0; j < largura_imagem; j++) {
            float soma = 0;

            for (int m = 0; m < altura_mascara; m++)
                for (int n = 0; n < largura_mascara; n++)
                {
                    int y = i - centro_altura + m;
                    int x = j - centro_largura + n;

                    if ((0 <= y && y < altura_imagem) && (0 <= x && x < largura_imagem))
                        soma += imagem_entrada.at<uchar>(y, x) * mascara.at<float>(m, n);
                }
            imagem_saida.at<uchar>(i, j) = (uchar) soma;           
        }
    
    return imagem_saida;
}

Mat gerar_mascara_media(const int dimensao) {
    Mat mascara = Mat::zeros(dimensao, dimensao, CV_32F); // 8 bits uchar 1 banda

    const float media = 1/(float)(dimensao * dimensao);

    for (int i = 0; i < dimensao; i++)
        for (int j = 0; j < dimensao; j++)
            mascara.at<float>(i, j) = media;

    return mascara; 
}